lib/multiplexer/multiplexer.go

```diff
@@ -246,44 +256,146 @@ func (m *Mux) detectAndForward(conn net.Conn) {
-	// before actual protocol traffic flows.
-	// if the first attempt encounters proxy it
-	// goes to the second pass to do protocol detection
+	// Before actual protocol traffic flows, we try to parse optional PROXY protocol headers,
+	// that can be injected by load balancers or our own proxies. There can be multiple PROXY
+	// headers. After they are parsed, last pass does the actual protocol detection itself.
+	// We allow only one unsigned PROXY header from external sources, if it's enabled, and unlimited
+	// amount (within maxPROXYPasses) of signed headers from our own proxies, which take precedence.
 	var proxyLine *ProxyLine
-	for i := 0; i < 2; i++ {
+	for i := 0; i < maxPROXYPasses; i++ {
 		proto, err := detectProto(reader)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 
 		switch proto {
 		case ProtoProxy:
-			if !enableProxyProtocol {
+			if !m.EnableExternalProxyProtocol {
 				return nil, trace.BadParameter("proxy protocol support is disabled")
 			}
+			// We allow only one unsigned proxy line
 			if proxyLine != nil {
 				return nil, trace.BadParameter("duplicate proxy line")
 			}
 			proxyLine, err = ReadProxyLine(reader)
 			if err != nil {
 				return nil, trace.Wrap(err)
 			}
+			// repeat the cycle to detect the protocol
 		case ProtoProxyV2:
-			if !enableProxyProtocol {
-				return nil, trace.BadParameter("proxy protocol support is disabled")
+			newProxyLine, err := ReadProxyLineV2(reader)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			verified := false
+			// If TLVs are empty we know it can't be signed, so we don't try to verify to avoid unnecessary load
+			if m.CertAuthoritiesGetter != nil && len(newProxyLine.TLVs) > 0 {
+				hostCACerts := [][]byte{}
+				certAuthorities, err := m.CertAuthoritiesGetter.GetCertAuthorities(m.context, types.HostCA, false)
+				if err != nil {
+					return nil, trace.Wrap(err, "could not get certificate authorities")
+				}
+
+				for _, ca := range certAuthorities {
+					caCerts := services.GetTLSCerts(ca)
+					hostCACerts = append(hostCACerts, caCerts...)
+				}
+				verified, err = newProxyLine.VerifySignature(hostCACerts, m.Clock)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			// If proxy line is signed and successfully verified, it's always accepted as current proxy line
+			if verified {
+				proxyLine = newProxyLine
+				continue
 			}
+
+			// If proxy line was signed but couldn't be verified it's ignored
+			if !verified && newProxyLine.isSigned() {
+				m.Warnf("Rejected signed PROXY header, incoming connection remote/local: %q, PROXY source/destination: %q",
+					conn.RemoteAddr().String()+"->"+conn.LocalAddr().String(),
+					newProxyLine.Source.String()+"->"+newProxyLine.Destination.String())
+				continue
+			}
+
+			// This is unsigned proxy line, return error if external PROXY protocol is not enabled
+			if !m.EnableExternalProxyProtocol {
+				return nil, trace.BadParameter("external proxy protocol support is disabled")
```

Comment:
We should have consistency here - if we are willing to ignore an invalid signed PROXYv2, we should probably ignore and log rather than reject here. (But as explained in a previous comment, I think we should switch to rejecting an invalid signed PROXYv2)

Reply:
Yep, we're now failing unsuccessfully verified signed header as well.